internal/monitors/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,414 @@
+	nodeIsCurrentMaster  bool
+	defaultDimensions    map[string]string
+	nodeMetricDimensions map[string]string
+	lock                 *sync.RWMutex
+}
+
+func (sinfo *sharedInfo) fetchNodeAndClusterMetadata(esClient client.ESHttpClient, configuredClusterName string) error {
+
+	// Collect info about master for the cluster
+	masterInfoOutput, err := esClient.GetMasterNodeInfo()
+
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to GET master node info")
+		return err
+	}
+
+	// Collect node info
+	nodeInfoOutput, err := esClient.GetNodeInfo()
+
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to GET node info")
+		return err
+	}
+
+	// Hold the lock while updating info shared between different fetchers
+	sinfo.lock.Lock()
+	defer sinfo.lock.Unlock()
+
+	nodeDimensions, err := prepareNodeMetricsDimensions(nodeInfoOutput.Nodes)
+
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to prepare node dimensions")
+		return err
+	}
+
+	sinfo.nodeIsCurrentMaster, err = isCurrentMaster(nodeDimensions["node_id"], masterInfoOutput)
+
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to determine whether current node is the master")
+		return err
+	}
+
+	clusterName := nodeInfoOutput.ClusterName
+	sinfo.defaultDimensions, err = prepareDefaultDimensions(configuredClusterName, clusterName)
+
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to prepare plugin_instance dimension")
+		return err
+	}
+
+	sinfo.nodeMetricDimensions = utils.MergeStringMaps(sinfo.defaultDimensions, nodeDimensions)
+
+	return nil
+}
+
+// Returns all fields of a shared info object
+func (sinfo *sharedInfo) getAllSharedInfo() (map[string]string, map[string]string, bool) {
+	sinfo.lock.RLock()
+	defer sinfo.lock.RUnlock()
+
+	return sinfo.defaultDimensions, sinfo.nodeMetricDimensions, sinfo.nodeIsCurrentMaster
+}
+
+// Configure monitor
+func (m *Monitor) Configure(conf *Config) error {
+	esClient := client.NewESClient(conf.Host, conf.Port, conf.UseHTTPS, conf.Username, conf.Password)
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
+	// To handle metadata about metrics that is shared
+	shared := sharedInfo{
+		lock:          &sync.RWMutex{},
+		isInitialized: false,
+	}
+
+	utils.RunOnInterval(m.ctx, func() {
+		// Fetch metadata from Elasticsearch nodes
+		err := shared.fetchNodeAndClusterMetadata(esClient, conf.Cluster)
+
+		// For any reason the monitor is not able to fetch cluster and node information upfront we capture that
+		// to ensure that the stats are not sent in with faulty dimensions. The monitor will try to fetch this
+		// information again every MetadataRefreshInterval seconds
+		if err != nil {
+			logger.WithError(err).Errorf(fmt.Sprintf("Failed to get Cluster and node metadata upfront. Will try again in %d s.", conf.MetadataRefreshIntervalSeconds))
+			return
+		}
+
+		// The first time fetchNodeAndClusterMetadata succeeded the monitor can schedule fetchers for all the other stats
+		if !shared.isInitialized {
```

Comment:
just put `isInitialized` as a variable in the main body of the `Configure` method instead of making it a member of `shared`.

Reply:
Done